REFRAME.md

```diff
@@ -0,0 +1,290 @@
+    }
+    
+    type GraphSync-FILv1 struct {
+        ID SingletonInt(0x0910) # https://github.com/multiformats/multicodec/blob/f5dd49f35b26b447aa380e9a491f195fd49d912c/table.csv#L134
+        DealID Bytes # TODO: Is there a spec for what goes here?
+    }
+```
+
+##### DAG-JSON Examples
+
+Request:
+```
+{"FindProviders" : {
+    "Key" : {"/":{"bytes":"AXIUBPnagss"}}
+}}
+```
+
+Response:
+```
+{"FindProviders" : {
+    "Key" : {"/":{"bytes":"AXIUBPnagss"}},
+    "Providers" : [
+        {"Node":{"Peer":{"ID":{
+            "/":{"bytes":"EncodedPeerID"}},
+            "Multiaddresses":[{"/":{"bytes":"EncodedAddr1"}}, {"/":{"bytes":"EncodedAddr2"}}]}}},
+        {"Node":{"Peer":{"ID":{
+            "/":{"bytes":"EncodedPeerID2"}},
+            "Multiaddresses":[{"/":{"bytes":"EncodedAddr1"}}, {"/":{"bytes":"EncodedAddr2"}}]}}},
+                "Proto" : [{
+                    "ID": 2320, # the integer of the graphsync-filv1 code
+                    "DealID" : {"/":{"bytes":"SomeDataIdentifier"}}
+                },
+                "Proto" : {
+                    "ID": 2304 # integer bitswap code
+                }]
+    ]
+}}
+```
+
+#### GetIPNS
+
+A message for finding the latest IPNS records for a given identifier.
+
+```ipldsch
+    type GetIPNSRequest struct {
+        ID Bytes # An IPNS Identifier, matches a peerID https://github.com/libp2p/specs/blob/f433ad595224cf33d916c166d1738f11aadfa9f7/peer-ids/peer-ids.md
+    }
+```
+
+```ipldsch
+    type GetIPNSResponse struct {
+        Record Bytes # An IPNS record
+    }
+```
+
+##### DAG-JSON Examples
+
+Request:
+```
+{"GetIPNS" : {
+    "ID" : {"/":{"bytes":"AXIUBPnagss"}}
+}}
+```
+
+Response:
+```
+{"GetIPNS" : {
+    "Record" : {"/":{"bytes":"firstRecord"}}
+}}{"GetIPNS" : {
+    "Record" : {"/":{"bytes":"laterRecord"}}
+}}...
+```
+
+#### PutIPNS
+
+A message for putting the latest IPNS records for a given identifier.
+
+```ipldsch
+    type PutIPNSRequest struct {
+        ID Bytes # IPNS Identifier
+        Record Bytes # IPNS record
+    }
+    
+    type PutIPNSResponse struct {}
+```
+
+##### DAG-JSON Examples
+
+Request:
+```
+{"PutIPNSRequest" : {
+    "ID" : {"/":{"bytes":"AXIUBPnagss"}},
+    "Record" : {"/":{"bytes":"AXIUBPnagss"}}
+}}
+```
+
+Response:
+```
+{"PutIPNSResponse : {}"}
+```
+
+# Method Upgrade Paths
+
+It is acknowledged that the initial methods and mechanisms of this protocol will likely need to change over time and that we should prepare for how to do so without the need to wholesale replace this protocol with an alternative.
+
+2. If it is desired to add optional parameters to a given method (in either a request or response) new fields can be added. It is recommended that new fields be submitted to this spec to prevent conflicts and for unsubmitted features to be added using namespaces unlikely to conflict (e.g. `MyProject-Field`).

```

Comment:
I don't know if within IPLD land we have any precedence for this type of namespacing we want to be consistent with.

In libp2p land we use paths (e.g. `/ipfs/bitswap/1.2.0`) and HTTP land has X Headers. Anything  different from `MyProject-...` we want to encourage here @mvdan?